File: src/evai/command/ExportCommands.java
package evai.command;

import mindustry.*;
import mindustry.entities.*;
import mindustry.gen.*;
import mindustry.world.*;
import arc.struct.*;
import java.util.*;

public class ExportCommands implements CommandRegistry.CommandHandler {
    @Override
    public CommandResult handle(String action, String[] args) {
        if (args.length == 0) return new CommandResult(false, "usage: export blocks/state/tile/units");
        switch (args[0].toLowerCase()) {
            case "blocks": return exportBlocks(args);
            case "state": return exportState();
            case "tile": return exportTile(args);
            case "units": return exportUnits(args);
            default: return new CommandResult(false, "unknown export type: " + args[0]);
        }
    }

    private CommandResult exportBlocks(String[] args) {
        String filter = args.length > 1 ? args[1] : "";
        Seq<Block> raw = Vars.content.blocks();
        List<Map<String, Object>> list = new ArrayList<>();
        for (int i = 0; i < raw.size; i++) {
            Block b = raw.get(i);
            if (filter.isEmpty() || b.name.contains(filter)) {
                Map<String, Object> m = new HashMap<>();
                m.put("name", b.name);
                m.put("type", b.getClass().getSimpleName());
                m.put("health", b.health);
                m.put("size", b.size);
                m.put("itemCapacity", b.itemCapacity);
                m.put("hasPower", b.hasPower);
                m.put("hasLiquids", b.hasLiquids);
                m.put("hasItems", b.hasItems);
                list.add(m);
            }
        }
        Map<String, Object> data = new HashMap<>();
        data.put("count", list.size());
        data.put("blocks", list);
        return new CommandResult(true, "", data);
    }

    private CommandResult exportState() {
        Map<String, Object> data = new HashMap<>();
        data.put("wave", (int) Vars.state.wave);
        data.put("map", Vars.state.map != null ? Vars.state.map.name() : "none");
        Map<String, Object> rules = new HashMap<>();
        rules.put("infiniteResources", Vars.state.rules.infiniteResources);
        rules.put("instantBuild", Vars.state.rules.instantBuild);
        rules.put("buildSpeedMultiplier", Vars.state.rules.buildSpeedMultiplier);
        data.put("rules", rules);
        return new CommandResult(true, "", data);
    }

    private CommandResult exportTile(String[] args) {
        if (args.length < 3) return new CommandResult(false, "usage: export tile <x> <y>");
        int x = tryParseInt(args[1], -1), y = tryParseInt(args[2], -1);
        if (x < 0 || y < 0) return new CommandResult(false, "invalid coordinates");
        Tile tile = Vars.world.tile(x, y);
        if (tile == null) return new CommandResult(false, "tile not found");
        Map<String, Object> data = new HashMap<>();
        data.put("x", tile.x); data.put("y", tile.y);
        data.put("block", tile.block() != null ? tile.block().name : "air");
        data.put("floor", tile.floor() != null ? tile.floor().name : "?");
        data.put("overlay", tile.overlay() != null ? tile.overlay().name : "none");
        data.put("team", tile.team() != null ? tile.team().name : "none");
        data.put("solid", tile.solid());
        return new CommandResult(true, "", data);
    }

    private CommandResult exportUnits(String[] args) {
        Player player = Vars.player;
        if (player == null) return new CommandResult(false, "no player");
        float radius = args.length > 1 ? tryParseFloat(args[1], 100f) : 100f;
        List<Map<String, Object>> list = new ArrayList<>();
        Units.nearby(player.x, player.y, radius * 2, radius * 2, u -> {
            Map<String, Object> m = new HashMap<>();
            m.put("name", u.type().name);
            m.put("health", (int) u.health() + "/" + (int) u.maxHealth());
            m.put("team", u.team().name);
            m.put("x", (int) u.x());
            m.put("y", (int) u.y());
            list.add(m);
        });
        Map<String, Object> data = new HashMap<>();
        data.put("count", list.size());
        data.put("units", list);
        return new CommandResult(true, "", data);
    }

    private static int tryParseInt(String s, int fallback) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return fallback; }
    }
    private static float tryParseFloat(String s, float fallback) {
        try { return Float.parseFloat(s); } catch (NumberFormatException e) { return fallback; }
    }
}

File: src/evai/command/ExportCommands.kt
package evai.command

import mindustry.*
import evai.command.CommandRegistry.CommandHandler

class ExportCommands : CommandHandler {
    override fun handle(action: String, args: Array<String>): CommandResult = when (action) {
        "export" -> {
            if (args.isEmpty()) return CommandResult(false, "usage: export blocks/state/tile/units [filter/radius]")
            when (args[0].lowercase()) {
                "blocks" -> {
                    val filter = if (args.size > 1) args[1] else ""
                    val blocks = Vars.content.blocks().filter { filter.isBlank() || it.name.contains(filter) }
                    CommandResult(true, data = mapOf("count" to blocks.size, "blocks" to blocks.map { mapOf("name" to it.name, "type" to it.javaClass.simpleName, "health" to it.health, "size" to it.size, "itemCapacity" to it.itemCapacity) }))
                }
                "state" -> {
                    val state = Vars.state
                    val rules = state.rules
                    CommandResult(true, data = mapOf("wave" to state.wave, "map" to (state.map?.name() ?: "none"), "rules" to mapOf("infiniteResources" to rules.infiniteResources, "instantBuild" to rules.instantBuild, "buildSpeedMultiplier" to rules.buildSpeedMultiplier)))
                }
                "tile" -> {
                    if (args.size < 3) return CommandResult(false, "usage: export tile <x> <y>")
                    val x = args[1].toIntOrNull() ?: return CommandResult(false, "invalid x")
                    val y = args[2].toIntOrNull() ?: return CommandResult(false, "invalid y")
                    val tile = Vars.world.tile(x, y) ?: return CommandResult(false, "tile not found")
                    CommandResult(true, data = mapOf("x" to tile.x, "y" to tile.y, "block" to (tile.block()?.name ?: "air"), "floor" to (tile.floor()?.name ?: "?"), "overlay" to (tile.overlay()?.name ?: "none"), "team" to (tile.team()?.name ?: "none"), "solid" to tile.solid()))
                }
                "units" -> {
                    val radius = args.getOrNull(1)?.toFloatOrNull() ?: 100f
                    val player = Vars.player ?: return CommandResult(false, "no player")
                    val list = mutableListOf<Map<String, Any?>>()
                    mindustry.entities.Units.nearby(player.x, player.y, radius) { u ->
                        list.add(mapOf("name" to (u.type?.name ?: "?"), "health" to "${u.health.toInt()}/${u.maxHealth.toInt()}", "team" to u.team.name, "x" to u.x.toInt(), "y" to u.y.toInt()))
                    }
                    CommandResult(true, data = mapOf("count" to list.size, "units" to list))
                }
                else -> CommandResult(false, "unknown export type: ${args[0]}")
            }
        }
        else -> CommandResult(false, "unknown export command: $action")
    }
}